third_party/move/move-compiler-v2/src/flow_insensitive_checkers.rs

```diff
@@ -0,0 +1,203 @@
+// Copyright © Aptos Foundation
+// SPDX-License-Identifier: Apache-2.0
+
+//! Flow-insensitive checks can be done on the AST.
+//!
+//! Warnings about Unused parameter and local variable
+//!   "Unused assignment or binding for local 's'. Consider removing, replacing with '_' or prefixing with '_' (e.g., '_r_ref')
+
+use codespan_reporting::diagnostic::Severity;
+use move_model::{
+    ast::{ExpData, TempIndex},
+    model::{GlobalEnv, Loc, NodeId, Parameter},
+    symbol::Symbol,
+};
+use std::{collections::BTreeSet, iter::Iterator};
+
+/// Warns about all parameters and local variables that are unused.
+pub fn check_for_unused_vars_and_params(env: &mut GlobalEnv) {
+    for module in env.get_modules() {
+        for func in module.get_functions() {
+            if let Some(def) = &*func.get_def() {
+                let params = &func.get_parameters();
+                find_unused_params_and_vars(env, params, def)
+            }
+        }
+    }
+}
+
+fn find_unused_params_and_vars(env: &GlobalEnv, params: &[Parameter], exp: &ExpData) {
+    let mut visitor = SymbolVisitor::new(env, params);
+    exp.visit_pre_post(&mut |post, exp_data| visitor.entry(post, exp_data));
+    visitor.check_parameter_usage();
+}
+
+/// Tracks things of type `V` which are visible from below in a tree, such as free/used variables,
+/// etc.  `values` is cleared when entering a scope, but the old value saved so it can be re-added
+/// once the scope is finished.
+struct ScopedVisibleSet<V> {
```

Comment:
nit: since this is nicely self-contained, it might be nice to add local unit tests.

Reply:
Please note that unit tests on this level have the tendency to also make code evolution harder. I'm not so convinced we need them, as for a compiler we can easily create e2e tests via source input. There are pros and cons.

Reply:
Yes, I agree it depends, and that's why I marked it as nit.

Having local unit tests for code like the symbol visitor, which is depending on the AST representation, is certainly brittle, and I do not see a high value proposition.

But for relatively well-contained data structures (and ones that may become more commonly used), having local unit tests might be nice because:
- it would help with testing any data structure optimizations (such as using immutable ones)
- it would prevent us from adding hacks that should not really be part of this data structure's abstractions and make it more general

All this to say, these being only potential benefits, I am totally okay with whatever the author's decision is here.